File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int res = 0;
        int l = 0;
        unordered_map<char, int> letMap;
        int maxletter = 0;

        for (int r = 0; r < s.size(); r++){
            letMap[s[r]]++;

            maxletter = max(maxletter, letMap[s[r]]);

            while ((r-l+1) - maxletter > k){
                letMap[s[l]]--;
                l++;
            }

            if ((r-l+1) > res){
                res = r-l+1;
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-4.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r, res = 1, max(piles), max(piles)

        while l <= r:
            k = (l + r) // 2

            time = 0
            for num in piles:
                time += math.ceil(num / k)
            if time <= h:
                res = k
                r = k-1
            else:
                l = k+1

        return res